src/components/networked-video-source-green-screen.js

```diff
@@ -0,0 +1,164 @@
+
+        this.uniforms = {};
+
+        this.uniforms.uMap = {type: 't', value: this.videoTexture }
+        this.uniforms.GreenThresholdIn = {type: 'float', value: this.data.GreenThresholdIn};
+
+
+
+        this.uniforms = THREE.UniformsUtils.merge([
+                                                      this.uniforms,
+                                                      THREE.UniformsLib['lights']
+                                                    ]);
+
+        this.materialIncoming =  new THREE.ShaderMaterial({
+                                  uniforms: this.uniforms
+                                });
+
+        this.materialIncoming.vertexShader = `
+                         varying vec2 vUv;
+
+                        void main() {
+                            vec4 worldPosition = modelViewMatrix * vec4( position, 1.0 );
+                            vec3 vWorldPosition = worldPosition.xyz;
+                            vUv = uv;
+                            gl_Position = projectionMatrix * modelViewMatrix * vec4( position, 1.0 );
+                        }
+                      `;
+
+        this.materialIncoming.fragmentShader = `
+                           varying vec2 vUv;
+                           uniform sampler2D uMap;
+                           uniform float GreenThresholdIn;
+                           
+                           void main() {
+                                vec2 uv = vUv;
+                                vec4 tex1 = texture2D(uMap, uv * 1.0);
+                                 if (tex1.g - tex1.r > GreenThresholdIn)
+                                    gl_FragColor = vec4(0,0,0,0);
+                                 else
+                                    gl_FragColor = vec4(tex1.r,tex1.g,tex1.b,1.0);
+                            }
+                      `;
+
+        this.materialIncoming.transparent = true;
+        this.materialIncoming.side = THREE.BackSide;
+        mesh.material = this.materialIncoming;
+
+        //---------- end of replace -----------------
+        //mesh.material.map = this.videoTexture;
+        mesh.material.needsUpdate = true;
+      }
+
+      this.stream = newStream;
+    }
+  },
+
+  _clearMediaStream() {
+
+    this.stream = null;
+
+    if (this.videoTexture) {
+
+      if (this.videoTexture.image instanceof HTMLVideoElement) {
+        // Note: this.videoTexture.image === this.video
+        const video = this.videoTexture.image;
+        video.pause();
+        video.srcObject = null;
+        video.load();
+      }
+
+      this.videoTexture.dispose();
+      this.videoTexture = null;
+    }
+  },
+
+  remove: function() {
+      this._clearMediaStream();
+  },
+
+  setupVideo: function() {
+    if (!this.video) {
+      const video = document.createElement('video');
+      video.setAttribute('autoplay', true);
+      video.setAttribute('playsinline', true);
+      video.setAttribute('muted', true);
+      this.video = video;
+    }
+  }
+});
```

Comment:
The idea I had was to make changes in the networked-video-source component, maybe adding a parameter or two to have the option of using a custom shader instead of writing a specific networked-video-source-green-screen, duplicating lots of the code.

Reply:
My mainly concern was not to break the existing component and have something for you to merge without conflict resolution. I can merge them if it is an issue.

Reply:
It doesn't make much sense to me to have two different components that are almost identical in the NAF library. It should be a properly registered shader and an option to the existing `networked-video-source` to use a custom shader. Of course the default options should have no change to the existing examples, so still using the default shader.
It depends on the level of contribution you are willing to take. If you decide to keep it separated, in this case you shouldn't include it in `src/components` but keep it only for your example, so in `examples/js` directory.

Reply:
Actually you shouldn't need to change networked-video-source at all. You should be able to register a new "green-screen" shader with `registerShader`, modify your shader to use uniform `map` instead of `uMap`. And use in your entity `networked-video-source="streamName: screen" material="shader:green-screen;GreenThresholdIn:0.02"`

Reply:
I have selected the last case, however,

*How do I update the map with the video texture ?*  I am completely stack with the shader below. My map is always null. I have not much experience with shaders and there only a few examples in Aframe.

```
<script>

    let vertexShader1 = `
                           varying vec2 vUv;

                          void main() {
                              vec4 worldPosition = modelViewMatrix * vec4( position, 1.0 );
                              vec3 vWorldPosition = worldPosition.xyz;
                              vUv = uv;
                              gl_Position = projectionMatrix * modelViewMatrix * vec4( position, 1.0 );
                          }
                        `;

    let fragmentShader1 = `
                             varying vec2 vUv;
                             uniform sampler2D map;
                             uniform float gthreshold;

                             void main() {
                                  vec2 uv = vUv;
                                  vec4 tex1 = texture2D(map, uv * 1.0);
                                  if (tex1.g - tex1.r > gthreshold)
                                      gl_FragColor = vec4(0,0,0,0);
                                  else
                                      gl_FragColor = vec4(tex1.r,tex1.g,tex1.b,1.0);
                              }`;


    AFRAME.registerShader('green-screen', {

        schema: {
            map: {type: 'map', is: 'uniform'},
            gthreshold : {type: 'float', value: 0.02}
        },

        uniforms: {
            // the texture value (once the texture source is loaded, update)
            map: { type: 't', value: null },
            gthreshold:  {type: 'f', value: 0.02}
        },

        init: function(data) {

            this.material = new THREE.ShaderMaterial({
                uniforms: this.uniforms,
                vertexShader: this.vertexShader,
                fragmentShader: this.fragmentShader
            });
        },

        update: function (data) {
            AFRAME.utils.material.updateMap(this, data);
            this.uniforms.map.value = this.material.map;
        },

        vertexShader: vertexShader1,
        fragmentShader: fragmentShader1
    });

</script>
```